### PR TITLE
earthgecko_skyline_detector

### PR DESCRIPTION
**Pull request by earthgecko**
Adding earthgecko_skyline detector code, results, etc.

### CODE REVIEW

**Comment by rhyolight** [comment]
@subutai If this is simply a better implementation of the skyline detector, is there any reason not to just use the version that performs better? I would prefer if we only had one version on the leaderboard. What do you think @earthgecko?

**Comment by subutai** [comment]
> @subutai If this is simply a better implementation of the skyline detector, is there any reason not to just use the version that performs better? I would prefer if we only had one version on the leaderboard. What do you think @earthgecko?

I think the main reason to keep the old implementation is that there have been published results with that version, so from a paper reproducibility standpoint it may be worth keeping. 

@earthgecko makes some really good points as to why that earlier implementation was insufficient, so I'm all for merging this version in (once any other review comments have been addressed).

**Comment by earthgecko** [comment]
Hi @rhyolight and @subutai there may be another option.  Seeing as the paper
references the v1.0 release, there should not be any harm in leaving the
original implementation out moving forward.  But it is just a matter of time and
energy to calculate the scores, so there is little harm in keeping it in future
versions either.

However I would suggest that if it is kept in, it is disabled by default and has
to be enabled to run.  That way it is not shipping/wasting time of any
unsuspecting/uninformed people that are using/testing with NAB.
I am happy enough to create a PR to enable that if you want to keep it in.

This may be especially true if more time series are added to the corpus as the
original implementation is always going to take relatively long to run,
especially on longer time series, even with the addition of the much improved
optimizer and scorer.

And on that note, I think that NAB may benefit from more real world time series
added to the corpus, which I shall endeavour to start collecting and submitting.
And I am pretty certain the numenta detector will score highest on additions
too ;)

**Comment by smirmik** [comment]
> And I am pretty certain the numenta detector will score highest on additions
> too ;)

There are different opinions about this ;)
Sorry for off topic.
Just wanted to remind about CAD OSE and that this is not the only version of CAD. ;)
@subutai , @rhyolight my regards. I miss the competition. You are great guys, but someone has to be on the dark (not opened) side ;)

**Comment by earthgecko** [comment]
@smirmik I think that at some point, we will all come to realise, that we are all actually on the same side

**Comment by earthgecko** [comment]
Wait until I add numenta and CAD OSE algorithms to Skyline and then we will see ;)
I could weight CONSENSUS values by NAB scores LOL

We are all anomaly hunters in an information dimension.   And to be honest, I am not sure that you lot and Jeff should be allowed to play with digital brains however you see fit :)

However it must be said that it is amazing what you have achieved to date with such a port, such an abstract endeavour!  I was glad to see that some of nupic scores have cyclonic complexity score of F, I would honesty expect nothing less.  Sincere kudos to the team, as always, a remarkable achievement.

@smirmik not to down play CAD OSE in anyway :)  It just that numenta must be acknowledged for having something something really, really unique and innovative on a much grander canvas then simple anomaly detection.

This took me much longer to post than expected, due to a github anomaly :)

There was an unexpected problem serving your request
Please try again and contact us if the problem persists including D26E:17271:1A1D6A5:2475A06:5C3F8F9A in your message

**Comment by rhyolight** [comment]
> I am happy enough to create a PR to enable that if you want to keep it in.

Would this be a change to this PR or are you talking about creating a new PR for it?

**Comment by earthgecko** [comment]
Hi @rhyolight it will be a separate PR.

File: nab/detectors/earthgecko_skyline/README.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+# CONSENSUS of 5 when the grubbs and ks_test algorithms are not enabled
+CONSENSUS = 5
+# CONSENSUS of 7 when the grubbs and ks_test algorithms are enabled
+# CONSENSUS = 7
+
+# Only use a sample of the data points of long time series
+SHORTEN_TIMESERIES = False
+# Based on 5 minute resolution data shorten to 7 days (513 data points) + 4 hrs
+# (60 data points) either side
+SHORTEN_TO_DATAPOINS = 633
```
Comment (rhyolight) [inline nab/detectors/earthgecko_skyline/README.md]:
```suggestion
SHORTEN_TO_DATAPOINTS = 633
```

**Comment by earthgecko** [comment]
@rhyolight I do not mind fixing all that actually.  If I have to look at it and work with it, it always bothers me when  things are not linted and it makes inline liniting in an IDE inefficient too.  I will take a look.

**Comment by rhyolight** [comment]
@subutai @earthgecko 

Aside from any outstanding code review issues....

Before I do anything at this point, I want to make it clear what we have agreed to. Here is the path forward I prefer, please review and comment before I take action:

1. We merge this PR, which adds "earthgecko skyline" to the leaderboard on the README
1. There will be no version bump
1. On version bump, we'll decide what to do about the two skyline implementations on the leaderboard.

If you guys agree to that, I approve this PR. 👍

**Comment by earthgecko** [comment]
@rhyolight sounds good to me.  If you would like any further changes with regards to aligning with the ```numenta/nupic/pylintrc``` please let me know which specifically and what you do not what changed, like the Skyline ```snake_case``` function names.

**Comment by subutai** [comment]
@rhyolight Sounds good to me! 

I assume we will later disable the old Skyline as the default case.

**Comment by earthgecko** [comment]
@rhyolight the funny thing is ...

```bash
(NAB-py2715-cp27mu) $ pylint --rcfile=$NUPIC/pylintrc nab/detectors/earthgecko_skyline/algorithms.py
Using config file ~/sandbox/of/github/numenta/nupic/pylintrc

------------------------------------
Your code has been rated at 10.00/10

(NAB-py2715-cp27mu) $ pylint --rcfile=$NUPIC/pylintrc nab/detectors/earthgecko_skyline/earthgecko_skyline_detector.py
Using config file ~/sandbox/of/github/numenta/nupic/pylintrc
************* Module nab.detectors.earthgecko_skyline.earthgecko_skyline_detector
C0412,  84: 2 - Imports from package nab are not grouped (ungrouped-imports)

-----------------------------------
Your code has been rated at 9.94/10

(NAB-py2715-cp27mu) $ pylint --rcfile=$NUPIC/pylintrc nab/detectors/earthgecko_skyline/skyline_algorithms.py
Using config file ~/sandbox/of/github/numenta/nupic/pylintrc

------------------------------------
Your code has been rated at 10.00/10

(NAB-py2715-cp27mu) $
```

I got it pylintrc'd for you :)  However I had to change skyline function names that were in ```snake_case``` to get those ratings, so it was not what you wanted any way ;)

**Pull request merged**